includes/collection/class-inbox.php

```diff
@@ -0,0 +1,140 @@
+		if ( \is_wp_error( $activity ) ) {
+			return $activity;
+		}
+
+		// Check for duplicate activity.
+		if ( ! \is_wp_error( self::get_by_guid( $activity->get_id() ) ) ) {
+			return false;
+		}
+
+		$title      = self::get_object_title( $activity->get_object() );
+		$visibility = is_activity_public( $activity ) ? ACTIVITYPUB_CONTENT_VISIBILITY_PUBLIC : ACTIVITYPUB_CONTENT_VISIBILITY_PRIVATE;
+
+		$inbox_item = array(
+			'post_type'    => self::POST_TYPE,
+			'post_title'   => sprintf(
+				/* translators: 1. Activity type, 2. Object Title or Excerpt */
+				__( '[%1$s] %2$s', 'activitypub' ),
+				$activity->get_type(),
+				\wp_trim_words( $title, 5 )
+			),
+			'post_content' => wp_slash( $activity->to_json() ),
+			// ensure that user ID is not below 0.
+			'post_author'  => \max( $user_id, 0 ),
+			'post_status'  => 'pending',
+			'guid'         => $activity->get_id(),
+			'meta_input'   => array(
+				'_activitypub_object_id'             => object_to_uri( $activity->get_object() ),
+				'_activitypub_activity_type'         => $activity->get_type(),
+				'_activitypub_activity_actor'        => Actors::get_type_by_id( $user_id ),
+				'_activitypub_activity_remote_actor' => object_to_uri( $activity->get_actor() ),
+				'activitypub_content_visibility'     => $visibility,
+			),
+		);
+
+		$has_kses = false !== \has_filter( 'content_save_pre', 'wp_filter_post_kses' );
+		if ( $has_kses ) {
+			// Prevent KSES from corrupting JSON in post_content.
+			\kses_remove_filters();
+		}
+
+		$id = \wp_insert_post( $inbox_item, true );
+
+		if ( $has_kses ) {
+			\kses_init_filters();
+		}
+
+		return $id;
+	}
+
+	/**
+	 * Get the title of an activity recursively.
+	 *
+	 * @param Base_Object $activity_object The activity object.
+	 *
+	 * @return string The title.
+	 */
+	private static function get_object_title( $activity_object ) {
+		if ( ! $activity_object ) {
+			return '';
+		}
+
+		if ( is_string( $activity_object ) ) {
+			$post_id = url_to_postid( $activity_object );
+
+			return $post_id ? get_the_title( $post_id ) : '';
+		}
+
+		// phpcs:ignore Universal.Operators.DisallowShortTernary.Found
+		$title = $activity_object->get_name() ?: $activity_object->get_content();
+
+		if ( ! $title && $activity_object->get_object() instanceof Base_Object ) {
+			// phpcs:ignore Universal.Operators.DisallowShortTernary.Found
+			$title = $activity_object->get_object()->get_name() ?: $activity_object->get_object()->get_content();
+		}
+
+		return $title;
+	}
+
+	/**
+	 * Get the inbox item by activity id.
+	 *
+	 * @param string $guid The activity id.
+	 *
+	 * @return int|null The inbox item id or null if not found.
+	 */
+	public static function get_by_guid( $guid ) {
+		global $wpdb;
+		// phpcs:ignore WordPress.DB.DirectDatabaseQuery.DirectQuery, WordPress.DB.DirectDatabaseQuery.NoCaching
+		$post_id = $wpdb->get_var(
+			$wpdb->prepare(
+				"SELECT ID FROM $wpdb->posts WHERE guid=%s AND post_type=%s",
+				\esc_url( $guid ),
```

Comment:
```suggestion
				\esc_sql( $guid ),
```
Not sure if `esc_sql()` is more appropriate here? Or `esc_url_raw()` or something?